text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,595 @@
+"Process" spans or other spans modelling processing operations.
+
+#### Instrumenting pull-based scenarios
+
+In pull-based consumer scenarios, the delivery of messages is requested by the
+application code. This usually involves a blocking call, which returns zero or
+more messages on completion.
+
+A "Receive" span covers such calls and should link to the "Create" or "Publish"
+spans of all messages that are forwarded via the respective call. Depending on
+the use case, "Receive" spans can correlate with "Process" spans or other spans
+modelling processing operations.
+
+#### General considerations for both push-based and pull-based scenarios
+
+The operations modelled by "Deliver" or "Receive" spans do not strictly refer
+to receiving the message from intermediaries, but instead refer to the
+application receiving messages for processing. If messages are fetched from the
+intermediary and forwarded to the application in one go, the whole operation
+might be covered by a "Deliver" or "Receive" span. However, libraries or SDKs
+might pre-fetch messages from intermediaries and cache those messages, and only
+forward messages to the application at a later time. In this case, the
+operation of pre-fetching and caching should not be covered by the "Deliver" or
+"Receive" spans.
+
+Operations covered by "Deliver" or "Receive" can forward zero messages (e. g.
+to notify the application that no message is available for processing), one
+message, or multiple messages (a batch of messages). "Deliver" and "Receive"
+spans should link to the "Create" or "Publish" span of the messages forwarded,
+thus those spans can link to zero, one, or multiple "Create" spans.
+
+If the operation covered by "Deliver" or "Receive" forwards exactly one
+message, in addition to creating a link to the "Create" span, the "Create" span
+may also be used as a parent of the "Deliver" or "Receive" span. While this is
+not possible for all scenarios (e. g. when receiving a batch of messages, or
+when parenting "Deliver" or "Receive" spans to an ambient context), it can
+improve the user experience in some scenarios.
+
+#### Settlement
+
+Messages can be settled in a variety of different ways. In some cases, messages
+are not settled at all (fire-and-forget), in other cases settlement operations
+are triggered manually by the user, and in callback scenarios settlement can be
+automatically triggered by messaging SDKs based on return values of callbacks.
+
+A "Settle" span should be created for every settlement operation, no matter if
+the settlement operation was manually triggered by the user or automatically
+triggered by SDKs. SDKs will in some cases auto-settle messages in
+push-scenarios when messages are delivered via callbacks. In cases where it is
+possible, it is recommended to create the "Settle" span in the scope of the
+"Deliver" span.
+
+Alternatively, an event can be created instead of a "Settle" span. Events could
```

Comment:
I probably forgot, but why? is there a need to have an event? why not span all the time? how would links work in this case?

Reply:
If I remember correctly, this was intended for reducing the number of created spans in some scenarios. There can be no links in this case.

Reply:
I'd like to revisit it and would suggest using spans unless there are stronger reasons not to. Rationale:
- deliver spans might be completed by the time settlement happens
- there might be no deliver span (when receive is done)
- there might be no ambient context
- spans don't seem to be much bigger than events in terms of payload, but better supported by tracing backends

Reply:
Discussed in the workgroup: mentions of events will be removed from the document.